### PR TITLE
feat(gatsby-source-shopify): download images option

### LINKED ISSUE #23724: Add option to disable downloading of images when using the `gatsby-source-shopify` plugin.
## Summary

I'm working on rebuilding an existing Shopify site as a Gatsby site using the `gatsby-source-shopify` plugin. It has ~600x items, and my internet connection isn't particularly good, so it can take a very long time for the site the build (even in develop mode). Sometimes the build will fail completely. If sharp tries to transform any of the images it takes even longer and is even less stable.

I think it might be a good idea to add an option to disable downloading the images, as we can still serve the images from Shopify's CDN. The Shopify CDN will let you resize images using the URL, so the performance is still pretty good. Here is a page with some info on doing this: <https://www.shopify.com/partners/blog/img-url-filter>

### Basic example

```js
plugins: [
  {
    resolve: `gatsby-source-shopify`,
    options: {
      shopName: process.env.SHOPIFY_SHOP_NAME,
      accessToken: accessToken: process.env.SHOPIFY_ACCESS_TOKEN,

      // Whether or not images should be downloaded locally so that they can be
      // transformed by sharp. You might want to disable this for stores with a
      // lot of products
      // Defaults to true
      downloadImage: false,
    },
  },
];
```

### Motivation

Adding this option will make it much easier for people to use this plugin if they have poor internet connections, less powerful computers, or are using it to connect to a store with a large number of images to download.

I plan to attempt to add this myself (along with the help of my friend @mrhut10), but I've never written/worked on a Gatsby plugin before, and have limited knowledge of GraphQL outside of using it within Gatsby.

I also wanted to make sure it sounded like a good idea first, and that I'm not overlooking anything before doing the work and submitting a PR only to have it rejected.

If this goes well, I want to also work on a companion plugin to use the images from the Shopify CDN in `gatsby-image` in much the same way as 
[gatsby-transformer-cloudinary](https://github.com/cloudinary-devs/gatsby-transformer-cloudinary)

I'm not sure if this is possible since we need a base64 image as well, but I'll open a seperate issue for that once I get some feedback on this issue.

### PR DESCRIPTION
**Pull request by mrhut10**
## Description
this PR aims to add a downloadImages option to the gatsby-source-plugin as described by @lukebennett88 in issue #23724
this prevents huge amounts of images being downloaded if you intended to use Shopify's CDN for images anyway greatly speeding up site build times.

This is my first PR to gatsby and neither @lukebennett88 or I feel it has had enough testing so hoping the community can test this implementation.

##Example Use
say you have a Shopify store with 500 products and 5 images per product. and you are happy to use Shopify's CDN (and transforming URLs) for the files. currently, the plugin will still download all of the images in this example that's 2500 images your downloading for no reason greatly slowing builds or timing out your build.

### Documentation
new Options to gatsby-shopify-plugin
- downloadImages: 
  - true (default) will download all shopify images
  - false: Shopify images won't download and in place local file will refer to a provided defaultImage ( so that graphQL types remain unchanged
- defaultImageUrl: default image url (if none provided will defaults to https://raw.githubusercontent.com/lukebennett88/gatsby-transformer-shopify-images/topic/shopify-option-downloadImages/plugins/gatsby-source-shopify/src/default.png

have changed the plugin readme file to reflect this @gatsbyjs/learning please read

## Related Issues
  should fix #23724 

NOTE:
in the file "packages/gatsby-source/shopify/src/nodes.js" in the method call "downloadImageAndCreateFileNode"
do I need to do any cache invalidation in this file?
I have some concern that the referencing to cache with the fileNodeID may mean that when downloadImages is set back to true (from previously being false) it may try to use the cached file rather than the shopify images.

### CODE REVIEW

**Comment by mrhut10** [comment]
Any advice anyone what I can do to have this reviewed sooner?

**Comment by beamercola** [comment]
Yes!! Been dreaming of this - same situation, many products (and the images are huge) so the whole app is failing right now and I can't boot it.

I'd love to test this - I'm having a hard time adding your branch into my project - do you have any pointers there?

**Comment by mrhut10** [comment]
> 
> 
> @mrhut10 Yes!! Been dreaming of this - same situation, many products (and the images are huge) so the whole app is failing right now and I can't boot it.
> 
> I'd love to test this - I'm having a hard time adding your branch into my project - do you have any pointers there?
> 
> edit: i ended up just plucking the shopify folder and pushing up to private repo as yarn dependency. this works perfectly - let's merge this bad boy i have like 4 projects i need this on!

@beamercola reallly glad to hear you got it working and your keen for it. 
thank @lukebennett88  for the idea which he talks about #23724 

next project @lukebennett88 and I want to implement is a graphQL fragment which is compatible with gatsby-image that uses Shopify's CDN. would you also be keen on that in your projects @beamercola ??

**Comment by lukebennett88** [comment]
@beamercola Glad you got it working! It's nice to see other people wanting this functionality as well, I think it could be very useful for a lot of people.

**Comment by beamercola** [comment]
@lukebennett88 Honestly I don't know how anybody does it without this.  I haven't come across a store that works yet.  I guess there's a lot of small 10 product stores out there.  Anyways great work, saved my ass today (it stopped deploying, 100+ failures).

**Comment by mrhut10** [comment]
@beamercola if any of your site's are open source, can I have a look at one that your using this feature? Regards

**Comment by beamercola** [comment]
@mrhut10 none are open source - but i'm happy to share with you personally!  hmu on twitter.  I made a `<ShopifyImage>` component that splits up the url and rebuilds based on props

**Comment by Haven3** [comment]
This is going to be a massive help to us. I'm really looking forward to it getting merged in! Thanks for all the work you guys have put into it!

**Comment by mrhut10** [comment]
@Haven3 your most welcome!!! I'm hoping too, Love advise how to get it reviewed faster.

and @beamercola love the sound of your ShopifyImage and loved to have a look at that!!! I'll try find you on twitter. My twitter is [ahut10](https://twitter.com/ahut10) and @lukebennett88 twitter is [luke_bennett_](https://twitter.com/luke_bennett_)

**Comment by mrhut10** [comment]
@pieh is there anything we can do to have this reviewed?

**Comment by smthomas** [comment]
@mrhut10 I just reviewed this and it looks good. Can you merge with the current version and fix the conflicts? Once those are fixed we can finally get this merged! Thanks for your contribution on this.

**Comment by mrhut10** [comment]
@smthomas gladly, I should get a chance to do so overnight (Australia time)
and yay

**Comment by wutangpaul** [comment]
> @smthomas gladly, I should get a chance to do so overnight (Australia time)
> and yay

Any chance of fixing these merge conflicts?

**Comment by mrhut10** [comment]
Glad there's a growing demand for this, sorry for the delay I'll look at and try resolve the merge conflicts tomorrow night.

:)

**Comment by mrhut10** [comment]
@wutangpaul @smthomas 

just dealt with the merge conflicts. was fairly straight forward merge, but i'm just setting up a demo project for myself now to make sure no surprises.

**Comment by ascorbic** [comment]
@mrhut10 Good stuff. It's currently failing tests because you need to run prettier over the code. Run `yarn format` in the root and it should fix it, then run `yarn lint` to check.

**Comment by mrhut10** [comment]
> @mrhut10 Good stuff. It's currently failing tests because you need to run prettier over the code. Run `yarn format` in the root and it should fix it, then run `yarn lint` to check.

cool thank you @ascorbic , I could see it falling but couldn't see where the details of the tests failing where.

as I'm using git sparse-checkout I'm just adding all of the config files I need to be able to run the relevant tests / formating.

**Comment by mrhut10** [comment]
@ascorbic looks like still some unit tests not passing, is this normal??
if not happy to address but wont be until tomorrow night now :)

**Comment by ascorbic** [comment]
@mrhut10 Yeah, these are unrelated. All PRs are failing right now.

**Comment by wutangpaul** [comment]
@mrhut10 Any luck with this? Do you need to re-trigger a build?

**Comment by mrhut10** [comment]
> @mrhut10 Any luck with this? Do you need to re-trigger a build?

yes please @wutangpaul I haven't tried to rerun the tests but when I did last time they where completely unrelated to my code changes.
been a while so there could be merge conflicts again.

**Comment by beamercola** [comment]
Anybody up to the task of updating this for newest version?  I would but I don't think I'm smart enough.  I'd even pay a bounty for this.

We gotta campaign to get this merged in - seems super important!

**Comment by mrhut10** [comment]
hay @beamercola I've brought up to date before but not sure who's buttons we need to press to get it merged.

to help our case to get merged, can you (and anyone else) describe the pain your having / problem this will solve.

and on my end, I will relook at bring up to date next week and being it has been a while sence I've played with I think I better put it though its passes testing it on a demo project again.

**Comment by beamercola** [comment]
@mrhut10 My painpoint is I'm managing ~5 Shopify stores on Gatsby, some of which have multiple hundred products (each with ~5 images per product).  Build time can be 20+ minutes.  The annoyance is that I'm not using the local files/sharp for Gatsby, I'm relying on the CDN.

The modifications you made were perfect, it saved my butt with some angry clients after I sold them on the promise of Gatsby.  But a quick look at the `gatsby-source-shopify` codebase it's looking very different from the initial fix and seems like this patch needs to be started all over.

I imagine there's a lot of Gatsby sites that include 5 products or so as a small "shop" section of the site.  But the force downloading of all images is limiting to those who are excited about replacing Shopify with Gatsby for e-commerce.

Thanks for your patch, it's still going strong but getting nervous as the codebase is diverging a bit and I'm unable to update to newer versions without open heart surgery

**Comment by mrhut10** [comment]
@beamercola  thank you so much for sharing your pain, sounds like your describing the exact circumstance that this branch / feature was made for when @lukebennett88 first thought it up.

 :) yay. 

but I haven't look much codebase with the changes they've recently done. I'll certainly try to have a look at it next week.
happy to walk you though what I find, and maybe an opportunity to pair program out a solution if you or anyone wanted to as well.

**Comment by KyleAMathews** [comment]
Hi all. So sorry this hasn't gotten in yet. It's a good change.

I tried merging master locally and it merged cleanly so it looks like this can get it easily still. @mrhut10 if you could give me write access to the PR, I can push the changes + do some testing and help get this in.

**Comment by mrhut10** [comment]
> Hi all. So sorry this hasn't gotten in yet. It's a good change.
> 
> I tried merging master locally and it merged cleanly so it looks like this can get it easily still. @mrhut10 if you could give me write access to the PR, I can push the changes + do some testing and help get this in.

Gladly and sounds exciting,
I'll get you the wire access to the repo tonight ( Australia time)

**Comment by mrhut10** [comment]
there you go @KyleAMathews I've added you as a collaborator on my fork of the project :)
thank you for looking at it.

**Comment by KyleAMathews** [comment]
Merged in master. We'll see how tests go.

File: packages/gatsby-source-shopify/README.md
Comment (KyleAMathews) [inline packages/gatsby-source-shopify/README.md:81]: why is this necessary? If you're not downloading images then we wouldn't want `localFile` on the schema correct cause there's nothing to query there?

File: packages/gatsby-source-shopify/src/nodes.js
Comment (KyleAMathews) [inline packages/gatsby-source-shopify/src/nodes.js:41]: we can just skip over this code if `downloadFiles` is set to false

File: packages/gatsby-source-shopify/src/gatsby-node.js
Comment (KyleAMathews) [inline packages/gatsby-source-shopify/src/gatsby-node.js:66]: I'm a bit iffy on this. Is there any big advantage over just passing an object around?

Comment (mrhut10) [reply packages/gatsby-source-shopify/src/gatsby-node.js:66]: yep thats cool, I can pull that out 👍

File: packages/gatsby-source-shopify/src/nodes.js
Comment (mrhut10) [reply packages/gatsby-source-shopify/src/nodes.js:41]: awesome, yeah not being hugely familiar with the code base I wasn't sure if I could just skip over it or if I would break something else where if a image wasn't provided, hence why I supplied a placeholder image in its place. which the cache would result in only one image being downloaded rather than all of the images.

thank you so much for pointing this out.

File: packages/gatsby-source-shopify/README.md
Comment (mrhut10) [reply packages/gatsby-source-shopify/README.md:81]: yeah I was swapping all of the images with a single placeholder image as wasn't sure if I would break something if a image wasn't supplied, but as per your comment on line `packages/gatsby-source-shopify/src/nodes.js:41` sounds like much better solution just to skip over a few lines with a nested if statement :)

**Comment by mrhut10** [comment]
thank you @KyleAMathews I'll let you know once I make those changes :)

**Comment by mrhut10** [comment]
just having a play with this now, also noticed in the future we might want to support turning off images for particular types of documents, i.e. Product Images, Collection Images, and Article Images if theres a needs for it could selectively turn this off. But I guess see if that gets requested after this base feature is in master.

File: packages/gatsby-source-shopify/README.md
Comment (mrhut10) [reply packages/gatsby-source-shopify/README.md]: sweet fixed this up,
you can see when downloadImages is set to false, true (default) then the localFile on the image node will or wont be in the graphQl types.

image

File: packages/gatsby-source-shopify/src/gatsby-node.js
Comment (mrhut10) [reply packages/gatsby-source-shopify/src/gatsby-node.js]: i've pulled that out now :)

**Comment by mrhut10** [comment]
Hay @KyleAMathews I've committed those changes from your review.

* Removed the pluginOptions singleton class
* Removal of defaultUrl or images
* if !downloadImages then early return (setting the localFile to undefined) before downloading the image

**Comment by mrhut10** [comment]
testing I've already done @KyleAMathews 

**Setup Environment**
* clone [gatsby store](https://github.com/gatsbyjs/store.gatsbyjs.org)
* resolve the gatsby-source-shopify plugin locally to my altered and build branch

**Test Cases**
1. not setting config.downloadImages ( then will default to being set to true, i.e. will download images)
2. setting config.downloadImages to false - removes the localFile from graphQl types on product images. 
image
3. setting config.downloadImages to true - adds the localFile from graphQl types on product images 
image

**Comment by mrhut10** [comment]
@beamercola / @gooball / @lukebennett88  are any of you already using shopify CDN for images?
and have you managed to still use the gatsby-image component or just using image components directly?

once this is merged thinking it would be good to look at the possibility to:
1. create a few blog posts showing people how to use the shopify cdn with this option to greatly decrease build times of sites.
2. potentially look at creating a gatsby-transformer-shopifyCdnImages which makes a gatsby-image compatible data, so people can still define the image sizes in the graphQL query, and get back a gatsby-image compatible data so that they can use the Component thats familiar and often recommended for Gatsby projects.

**Comment by lukebennett88** [comment]
Hey @mrhut10 I haven't looked into this for a while.

To resize an image I would append `_${width}x` just before the file extension.

e.g.:

Original:
`https://cdn.shopify.com/s/files/1/0513/3785/5157/products/2711.jpg`

Resized:
`https://cdn.shopify.com/s/files/1/0513/3785/5157/products/2711_200x.jpg`

I wrote a crude function to resize them which was good enough for how I was using it:
```js
function resizeShopifyImage({ url, width }) {
  let fileExtension = '.jpg';
  let index = url.indexOf('.jpg');
  if (index === -1) {
    index = url.indexOf('.png');
    fileExtension = '.png';
  }
  if (index !== -1) {
    const firstChunk = url.slice(0, index);
    return firstChunk.concat(`_${width}x${fileExtension}`);
  }
  return url;
}

export { resizeShopifyImage };
```

I know at the time I there was an issue with the CDN — if I requested an image that was larger that the original image, the image wouldn't load. This doesn't seem to be the case anymore so a using this with Gatsby Image should probably work (I'm fairly certain that the base64 image isn't required since there is a Gatsby Image fragment with no base64 encoded image:
https://www.gatsbyjs.com/plugins/gatsby-image/?=gatsby-image#gatsby-transformer-sharp

One other thing to consider is that the Gatsby Image team are working on a replacement for Gatsby Image — [Gatsby Plugin Image](https://www.gatsbyjs.com/plugins/gatsby-plugin-image/) so it's probably a good idea to make it compatible with that so we don't have to redo it later.

**Comment by KyleAMathews** [comment]
>One other thing to consider is that the Gatsby Image team are working on a replacement for Gatsby Image — Gatsby Plugin Image so it's probably a good idea to make it compatible with that so we don't have to redo it later.

👍  to this. Here's an example of how source plugins can do the integration — https://github.com/gatsbyjs/gatsby/pull/28236

**Comment by mrhut10** [comment]
@KyleAMathews is there anyone we can request from gatsbyjs/documentation to review?

**Review by KyleAMathews — APPROVED** [review]
Thanks! Sorry again for the slow merge.

**Comment by KyleAMathews** [comment]
I've got the power ;-)

**Comment by mrhut10** [comment]
Yay!!!! so excited!! thank you @KyleAMathews 

@beamercola and/or @lukebennett88 can I use one of your sites that has too many shopify images in a blog post showing the build time performance increase? or/and interested in writing one yourself? (I still need to set up a blog or medium or something)

**Comment by beamercola** [comment]
@KyleAMathews for prez

@mrhut10 yeah, happy to!

**Comment by lukebennett88** [comment]
> Yay!!!! so excited!! thank you @KyleAMathews
> 
> @beamercola and/or @lukebennett88 can I use one of your sites that has too many shopify images in a blog post showing the build time performance increase? or/and interested in writing one yourself? (I still need to set up a blog or medium or something)

So happy to see this merged!
I'll update the source plugin on [GLF Online](https://www.glfonline.com.au/) tomorrow and most the results.

**Comment by lukebennett88** [comment]
How do the npm releases work?
Are they updated on a schedule? Hoping to test this tomorrow if it's released by then 🤞

**Comment by mrhut10** [comment]
@lukebennett88 its possible its automated, but otherwise I believe they would have to increment the version number in the package.json and then do a deployment to npm.

in the meanwhile to run locally in your dev environment

1. could copy the files from gatsby repo `packages/gatsby-source-shopify` folder into your project under the folder `/packages/gatsby-source-shopify`
2.  then cd into that folder and run `npm install && npm build`
3. then update your gatsby-config.json file to reflect that you want to resolve the plugin locally
```
// gatsby-config.js
{
      resolve: require.resolve(`./packages/gatsby-source-shopify`),
      options: {
        shopName: 'gatsby-swag',
        accessToken: process.env.SHOPIFY_ACCESS_TOKEN,
        downloadImages: true
      }
 },
```


If you want to run in production, you will need to do 
1. `git add` and commit the files in `packages/gatsby-source-shopify` (not the `gitignore` file they provide should already ignore the files that the build command generates.
2.  update your build command to also build the package i.e. in your package.json scripts `"build": cd packages/gatsby-source-shopify && npm install && npm build && cd ../.. && gatsby build"` (dont quote me on that one but will give you the idea hopefully

**Comment by lukebennett88** [comment]
Thanks @mrhut10, yeah I know I could get it working manually, but didn't want to go through the extra work if the package is likely to be updated soon.

Appreciate the write-up though, and thanks for going the last mile to finally get this thing merged in. So happy 🎉

**Comment by mrhut10** [comment]
@lukebennett88 looks like its been tagged and deployed under the tag [3.9.0-next.1](https://www.npmjs.com/package/gatsby-source-shopify/v/3.9.0-next.1) so you should be able to npm install that tag and use this feature with that version tag.
```
//npm
npm install gatsby-source-shopify@3.9.0-next.1
// yarn
yarn add gatsby-source-shopify@3.9.0-next.1
```


likewise you could always ask @beamercola to redeploy his npm version which he was using for his sites.
[npm link](https://www.npmjs.com/package/beamercola-gatsby-source-shopify)

**Pull request merged**
